### DISCUSSION TITLE
AIR SDK inside a docker image

### DISCUSSION THREAD

**Discussion by j3k0** [airsdk/Adobe-Runtime-Support#1254] [Ideas]
A great while back, I packaged the [AIR SDK inside a docker image](https://hub.docker.com/r/jeko/airbuild) so AIR's build tools can be ran from any computer having only docker installed. It removed a lot of headaches, like building from a CI server and setting up new computers for development.

At the time, the trick was to run AIR for Windows inside a Docker container using [Wine](https://www.winehq.org/).

I tried recently to update this image to use the AIR SDK for Linux, but the tools require a full working X environment and I failed to have this work at the moment using Xvfb inside the docker image.

Anyway, for us that is a great way to get a reproducible development environment: a docker image packages a well defined and compatible combination of AIR SDK, Java and Android SDK. The exact same build environment can be used from any computer running docker, regardless of the OS. (Windows, Linux and macOS)

**Comment by ajwfrost** [comment]
While I think it's a great idea to be able to set up the SDK and build tools within a docker container, I would have to point out the licensing challenges with this. You have pointed people to the license agreement that they would need to agree to, but your provision of the docker files presumably means you're breaking the "No Distribution of Build Tools" clause of that agreement - the link isn't working but I'm looking at the license PDF that's within Adobe's AIR SDK 32, clause 3.3(a)(iv).

I'd love to simplify the creation of the environment - and am delighted to see the [AIR Package Manager](https://github.com/airsdk/apm) development which is going to be of great help here. We had wondered about trying to remove the dependence upon Java, although for Android this is just the standard tool due to their use of Gradle etc - I'm not sure we'd be able to distribute the Java runtime itself (old versions perhaps - although our lawyers would likely baulk at the indemnification requirements - but the newer versions require commercial licenses in order to redistribute the runtime I believe).

We can take a look at this, it may be that we can create some cut-down version of the SDK which is just the build tools within a container in order to be used  for CI purposes but it may also need other dependencies so may end up just being a set of instructions to create your own..


thanks

**Reply by j3k0** [reply]
For us, at least, that'd be useful.

The docker image can be an extension of the official OpenJDK image: https://hub.docker.com/_/openjdk and I believe there's no issues with distributing the OpenJDK.

As for the non-open JDK, Oracle also distributes in on Docker Hub with the warning:

> Please review the terms carefully before downloading and using this product

https://hub.docker.com/_/oracle-jdk

If that's fine with Oracle to do it, it might be OK for your lawyers ;-)

**Reply by marchbold** [reply]
This is a great idea, would love to see something like this moving forward. Automating tests and builds would become a lot easier and allow simple usage in CI tools.

**Reply by AdamHarte** [reply]
How is setting up a local docker container any different from setting up your build environment locally outside of a container. Seems like an excuse, or a missunderstanding of how docker works. You are not distributing anything right? Except for configs about how the container should set itself up. Setting up everything needed to build an AIR app can be a massive pain. So being able to define a docker container would be amazing.

**Reply by marchbold** [reply]
I suppose the difference is the acceptance of the license, Docker would automate this and potentially violate something there? Probably a way around it in making the license part of the docker image usage terms or something?
